src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,163 @@
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  return self->state == NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+static lf_ret_t UARTPollChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+
+  if (self->state == NETWORK_CHANNEL_STATE_CONNECTED) {
+    int message_size = serialize_to_protobuf(message, self->write_buffer, UART_CHANNEL_BUFFERSIZE);
+
+    UART_CHANNEL_DEBUG("Sending Message of Size: %i", message_size);
+
+    uart_write(self->uart_dev, self->write_buffer, message_size);
+    return LF_OK;
+  } else {
+    return LF_ERR;
+  }
+}
+
+static void UARTPollChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                      void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                               const FederateMessage *msg),
+                                                      FederatedConnectionBundle *conn) {
+  UART_CHANNEL_INFO("Register receive callback");
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+void _UARTPollChannel_interrupt_callback(void *arg, uint8_t received_byte) {
+  UARTPollChannel *self = (UARTPollChannel *)arg;
+  const uint32_t minimum_message_size = 12;
+
+  self->receive_buffer[self->receive_buffer_index] = received_byte;
+  self->receive_buffer_index++;
+
+  if (self->receive_buffer_index >= minimum_message_size) {
+    if (self->super.super.mode == NETWORK_CHANNEL_MODE_ASYNC) {
+      cond_signal(&((UARTAsyncChannel *)self)->receive_cv);
+    }
+  }
+}
+
+void UARTPollChannel_poll(NetworkChannel *untyped_self) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  const uint32_t minimum_message_size = 12;
+
+  while (self->receive_buffer_index > minimum_message_size) {
+    int bytes_left = deserialize_from_protobuf(&self->output, self->receive_buffer, self->receive_buffer_index);
+    UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+    if (bytes_left >= 0) {
+      int receive_buffer_index = self->receive_buffer_index;
+      self->receive_buffer_index = bytes_left;
+      memcpy(self->receive_buffer, self->receive_buffer + (receive_buffer_index - bytes_left), bytes_left);
+
+      if (self->receive_callback != NULL) {
+        UART_CHANNEL_DEBUG("calling user callback!");
+        self->receive_callback(self->federated_connection, &self->output);
+      }
+    }
+  }
+}
+
+void *_UARTAsyncChannel_decode_loop(void *arg) {
+  UARTAsyncChannel *self = (UARTAsyncChannel *)arg;
+  mutex_lock(&self->receive_lock);
+
+  UART_CHANNEL_DEBUG("Entering decoding loop");
+  while (true) {
+    cond_wait(&self->receive_cv, &self->receive_lock);
+
+    UARTPollChannel_poll((NetworkChannel *)&self->super);
```

Comment:
I think we have to disable IRQs before calling `UARTPollChannel_poll` here. Even if you hold the `receive_lock`, I dont think there is anything stopping an interrupt from arriving and start messing around with the receive buffer while you are in the middle of a copy.

Reply:
incrementing overwriting the `receive_buffer_index` is now protected.